apis/apps/v1/componentdefinition_types.go

```diff
@@ -0,0 +1,1973 @@
+
+	// Port references a port or node-port defined in the service.
+	//
+	// If the referenced service is a pod-service, there will be multiple service objects matched,
+	// and the value will be presented in the following format: service1.name:port1,service2.name:port2...
+	//
+	// +optional
+	Port *NamedVar `json:"port,omitempty"`
+}
+
+// CredentialVars defines the vars that can be referenced from a Credential (SystemAccount).
+// !!!!! CredentialVars will only be used as environment variables for Pods & Actions, and will not be used to render the templates.
+type CredentialVars struct {
+	// +optional
+	Username *VarOption `json:"username,omitempty"`
+
+	// +optional
+	Password *VarOption `json:"password,omitempty"`
+}
+
+// ServiceRefVars defines the vars that can be referenced from a ServiceRef.
+type ServiceRefVars struct {
+	// +optional
+	Endpoint *VarOption `json:"endpoint,omitempty"`
+
+	// +optional
+	Host *VarOption `json:"host,omitempty"`
+
+	// +optional
+	Port *VarOption `json:"port,omitempty"`
+
+	CredentialVars `json:",inline"`
+}
+
+// HostNetworkVarSelector selects a var from host-network resources.
+type HostNetworkVarSelector struct {
+	// The component to select from.
+	ClusterObjectReference `json:",inline"`
+
+	HostNetworkVars `json:",inline"`
+}
+
+// ServiceVarSelector selects a var from a Service.
+type ServiceVarSelector struct {
+	// The Service to select from.
+	// It can be referenced from the default headless service by setting the name to "headless".
+	ClusterObjectReference `json:",inline"`
+
+	ServiceVars `json:",inline"`
+}
+
+// CredentialVarSelector selects a var from a Credential (SystemAccount).
+type CredentialVarSelector struct {
+	// The Credential (SystemAccount) to select from.
+	ClusterObjectReference `json:",inline"`
+
+	CredentialVars `json:",inline"`
+}
+
+// ServiceRefVarSelector selects a var from a ServiceRefDeclaration.
+type ServiceRefVarSelector struct {
+	// The ServiceRefDeclaration to select from.
+	ClusterObjectReference `json:",inline"`
+
+	ServiceRefVars `json:",inline"`
+}
+
+// ComponentVarSelector selects a var from a Component.
+type ComponentVarSelector struct {
+	// The Component to select from.
+	ClusterObjectReference `json:",inline"`
+
+	ComponentVars `json:",inline"`
+}
+
+type ComponentVars struct {
+	// Reference to the name of the Component object.
+	//
+	// +optional
+	ComponentName *VarOption `json:"componentName,omitempty"`
+
+	// Reference to the replicas of the component.
+	//
+	// +optional
+	Replicas *VarOption `json:"replicas,omitempty"`
+
+	// Reference to the instanceName list of the component.
+	// and the value will be presented in the following format: instanceName1,instanceName2,...
+	//
+	// +optional
+	InstanceNames *VarOption `json:"instanceNames,omitempty"`
+
+	// Reference to the pod FQDN list of the component.
+	// The value will be presented in the following format: FQDN1,FQDN2,...
+	//
+	// +optional
+	PodFQDNs *VarOption `json:"podFQDNs,omitempty"`
```

Comment:
Currently, except the load balancer IP, the provided addresses (e.g., pod FQDN, service host) are mostly DNS domain names or parts of them. 

There should be a way to provide a list of node hostnames, pod IPs, or service IPs that disclose the components' pods reside on. This way, in scenarios like hostNetwork or when there is no service (e.g., Kuaishou), IP addresses can be constructed accordingly.

Node or pod IPs cannot be obtained during the provisioning stage, so dynamic modification and notification for the resolution of vars values need to be supported first.